### Luis Eduardo Jiménez del Muro   -   05/09/2023

# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

#### 1. Criterio de optimización:
Minimización de costos de la producción del alimento que cumpla con las necesidades del gato.

#### 2. Función de costos:
$$ C = 0.013x_1 + 0.008x_2 + 0.010x_3 + 0.002x_4 + 0.005x_5 + 0.001x_6 $$

#### 3. Restricciones:
- **Mínimo de proteína:**
$ 0.10𝑥_1 + 0.20𝑥_2 + 0.15𝑥_3 + 0.00𝑥_4 + 0.04𝑥_5 + 0.00𝑥_6 ≥ 8 $

- **Mínimo de grasa:**
$ 0.08𝑥_1 + 0.10𝑥_2 + 0.11𝑥_3 + 0.01𝑥_4 + 0.01𝑥_5 + 0.00𝑥_6 ≥ 6 $

- **Fibra máxima**
$ 0.001𝑥_1 + 0.005𝑥_2 + 0.005𝑥_3 + 0.10𝑥_4 + 0.15𝑥_5 + 0.00𝑥_6 ≤ 2 $

- **Sal máxima**
$ 0.002𝑥_1 + 0.005𝑥_2 + 0.007𝑥_3 + 0.002𝑥_4 + 0.008𝑥_5 + 0.00𝑥_6 ≤ 0.4 $

#### 3. Restricciones (cambiando signo para que todos sean menor igual):
- **Mínimo de proteína:**
$ -0.10𝑥_1 - 0.20𝑥_2 - 0.15𝑥_3 - 0.00𝑥_4 - 0.04𝑥_5 - 0.00𝑥_6 ≤ -8 $

- **Mínimo de grasa:**
$ -0.08𝑥_1 - 0.10𝑥_2 - 0.11𝑥_3 - 0.01𝑥_4 - 0.01𝑥_5 - 0.00𝑥_6 ≤ -6 $

- **Fibra máxima**
$ 0.001𝑥_1 + 0.005𝑥_2 + 0.005𝑥_3 + 0.10𝑥_4 + 0.15𝑥_5 + 0.00𝑥_6 ≤ 2 $

- **Sal máxima**
$ 0.002𝑥_1 + 0.005𝑥_2 + 0.007𝑥_3 + 0.002𝑥_4 + 0.008𝑥_5 + 0.00𝑥_6 ≤ 0.4 $

- **Cantidad de comida**
$ x_1 + x_2 + x_3 + x_4 + x_5 + x_6 = 100 $

#### 4. Solución en código

In [11]:
import scipy.optimize as opt
import numpy as np

c = np.array([0.013,
              0.008,
              0.010,
              0.002,
              0.005,
              0.001])

Aub = np.array([
    [-0.10, -0.20, -0.15, -0.00, -0.04, -0.00],
    [-0.08, -0.10, -0.11, -0.01, -0.01, -0.00],
    [0.001, 0.005, 0.005, 0.10, 0.15, 0.00],
    [0.002, 0.005, 0.007, 0.002, 0.008, 0.00]])

bub = np.array([-8,
                -6,
                2,
                0.4])

Aeq = np.array([
    [1, 1, 1, 1, 1, 1]
])

beq = np.array([100])


sol = opt.linprog(c, Aub, bub, Aeq, beq)
print("Costo mínimo total:", sol.fun)
print("Cantidad de gramos de uso por ingrediente: ", "\n", sol.x)

Costo mínimo total: 0.52
Cantidad de gramos de uso por ingrediente:  
 [ 0. 60.  0.  0.  0. 40.]


In [12]:
Aub @ sol.x

array([-12. ,  -6. ,   0.3,   0.3])

- Por lo tanto, el costo mínimo de la comida es de $0.52$ y llevaría $60$ gramos de cordero y $40$ gramos de gel para no pasar las restricciones.

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

#### 1. Criterio de optimización:
Minimización de costos de la distribución de energía

#### 2. Función de costos:
$$ C = 5x_1 + 2x_2 + 7x_3 + 3x_4 + 6x_5 + 6x_6 + 6x_7 + x_8 + 2x_9 $$

#### 3. Restricciones:
- **Distribución de las plantas:**
    1. $ x_1 + x_2 + x_3 ≤ 80 $
    2. $ x_4 + x_5 + x_6 ≤ 40 $
    3. $ x_7 + x_8 + x_9 ≤ 60 $

- **Necesidad de las ciudades:**
    1. $ x_1 + x_4 + x_7 = 70 $
    2. $ x_2 + x_5 + x_8 = 40 $
    3. $ x_3 + x_6 + x_9 = 70 $

#### 4. Solución en código

In [13]:
import scipy.optimize as opt
import numpy as np

c = np.array([5,
              2,
              7,
              3,
              6,
              6,
              6,
              1,
              2])

Aub = np.array([
    [1, 1, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 1, 1, 1, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 1, 1, 1]
])

bub = np.array([80,
                40,
                60,])

Aeq = np.array([
    [1, 0, 0, 1, 0, 0, 1, 0, 0],
    [0, 1, 0, 0, 1, 0, 0, 1, 0],
    [0, 0, 1, 0, 0, 1, 0, 0, 1]
])

beq = np.array([70,
                40,
                70])

sol = opt.linprog(c, Aub, bub, Aeq, beq)
print("Costo mínimo total:", sol.fun)
print("Distribución óptima a las plantas de millones de Kw:", "\n", sol.x)

Costo mínimo total: 540.0
Distribución óptima a las plantas de millones de Kw: 
 [30. 40. 10. 40.  0.  0.  0.  0. 60.]


In [14]:
Aub @ sol.x

array([80., 40., 60.])

Por lo tanto, el costo mínimo de la distribución es de $540$ y se distribuirían de la siguiente manera:
1. **Guadalajara**
- 30 millones de kw a la planta 1, 40 millones de kw a la planta 2
2. **León**
- 40 millones de kw a la planta 1
3. **Morelia**
- 10 millones de kw a la planta 1, 60 millones de kw a la planta 3
